File: tui/src/scala/tui/widgets/SparklineWidget.scala
package tui
package widgets

import tui.internal.ranges

/** Widget to render a sparkline over one or more lines.
  *
  * @param block
  *   A block to wrap the widget in
  * @param style
  *   Widget style
  * @param data
  *   A slice of the data to display
  * @param max
  *   The maximum value to take to compute the maximum bar height (if nothing is specified, the widget uses the max of the dataset)
  * @param barSet
  *   A set of bar symbols used to represent the give data
  */
case class SparklineWidget(
    style: Style = Style.DEFAULT,
    data: collection.Seq[Int] = Nil,
    max: Option[Int] = None,
    barSet: symbols.bar.Set = symbols.bar.NINE_LEVELS
) extends Widget {
  def render(area: Rect, buf: Buffer): Unit = {
    if (area.height < 1) {
      return
    }

    val max = this.max match {
      case Some(v) => v
      case None    => this.data.maxOption.getOrElse(1)
    }
    val max_index = math.min(area.width, this.data.length)
    val data = this.data.take(max_index).toArray.map { e =>
      if (max != 0) {
        e * area.height * 8 / max
      } else {
        0
      }
    }

    ranges.revRange(0, area.height) { j =>
      ranges.range(0, data.length) { i =>
        val d = data(i)
        val symbol = d match {
          case 0 => barSet.empty
          case 1 => barSet.oneEighth
          case 2 => barSet.oneQuarter
          case 3 => barSet.threeEighths
          case 4 => barSet.half
          case 5 => barSet.fiveEighths
          case 6 => barSet.threeQuarters
          case 7 => barSet.sevenEighths
          case _ => barSet.full
        }
        buf
          .get(area.left + i, area.top + j)
          .setSymbol(symbol)
          .setStyle(style)

        if (d > 8) {
          data(i) -= 8
        } else {
          data(i) = 0
        }
      }
    }
  }
}

File: tui/src/java/tui/widgets/SparklineWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Buffer;
import tui.Rect;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;

/// Widget to render a sparkline over one or more lines.
public final class SparklineWidget implements Widget {
  public final Optional<BlockWidget> block;
  public final Style style;
  public final int[] data;
  public final Optional<Integer> max;
  public final Symbols.bar.Set barSet;

  public SparklineWidget(
      Optional<BlockWidget> block,
      Style style,
      int[] data,
      Optional<Integer> max,
      Symbols.bar.Set barSet) {
    this.block = block;
    this.style = style;
    this.data = data;
    this.max = max;
    this.barSet = barSet;
  }

  public static SparklineWidget empty() {
    return new SparklineWidget(
        Optional.empty(), Style.DEFAULT, new int[0], Optional.empty(), Symbols.bar.NINE_LEVELS);
  }

  public SparklineWidget withBlock(BlockWidget b) {
    return new SparklineWidget(Optional.of(b), style, data, max, barSet);
  }

  public SparklineWidget withStyle(Style s) {
    return new SparklineWidget(block, s, data, max, barSet);
  }

  public SparklineWidget withData(int[] d) {
    return new SparklineWidget(block, style, d, max, barSet);
  }

  public SparklineWidget withMax(int m) {
    return new SparklineWidget(block, style, data, Optional.of(m), barSet);
  }

  public SparklineWidget withBarSet(Symbols.bar.Set s) {
    return new SparklineWidget(block, style, data, max, s);
  }

  @Override
  public void render(Rect area, Buffer buf) {
    Rect sparkArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      sparkArea = innerArea;
    } else {
      sparkArea = area;
    }

    if (sparkArea.height() < 1) {
      return;
    }

    int maxValue;
    if (max.isPresent()) {
      maxValue = max.get();
    } else {
      int m = 1;
      boolean any = false;
      for (int v : data) {
        if (!any || v > m) {
          m = v;
          any = true;
        }
      }
      maxValue = any ? m : 1;
    }
    int maxIndex = Math.min(sparkArea.width(), data.length);
    int[] data2 = new int[maxIndex];
    for (int i = 0; i < maxIndex; i++) {
      if (maxValue != 0) {
        data2[i] = data[i] * sparkArea.height() * 8 / maxValue;
      } else {
        data2[i] = 0;
      }
    }

    Ranges.revRange(
        0,
        sparkArea.height(),
        j -> Ranges.range(
            0,
            data2.length,
            i -> {
              int d = data2[i];
              String symbol =
                  switch (d) {
                    case 0 -> barSet.empty();
                    case 1 -> barSet.oneEighth();
                    case 2 -> barSet.oneQuarter();
                    case 3 -> barSet.threeEighths();
                    case 4 -> barSet.half();
                    case 5 -> barSet.fiveEighths();
                    case 6 -> barSet.threeQuarters();
                    case 7 -> barSet.sevenEighths();
                    default -> barSet.full();
                  };
              buf.get(sparkArea.left() + i, sparkArea.top() + j)
                  .setSymbol(symbol)
                  .setStyle(style);

              if (d > 8) {
                data2[i] -= 8;
              } else {
                data2[i] = 0;
              }
            }));
  }
}
